test/test_lib.py

```diff
@@ -0,0 +1,442 @@
+            ops.lib.use('foo', 2, ops.lib.use)
+
+    def test_use_bad_args_values(self):
+        with self.assertRaises(ValueError):
+            ops.lib.use('--help', 2, 'alice@example.com')
+        with self.assertRaises(ValueError):
+            ops.lib.use('foo', -2, 'alice@example.com')
+        with self.assertRaises(ValueError):
+            ops.lib.use('foo', 1, 'example.com')
+
+
+@patch('sys.path', new=())
+class TestLibFunctional(TestCase):
+
+    def _mkdtemp(self) -> str:
+        tmpdir = mkdtemp()
+        self.addCleanup(rmtree, tmpdir)
+        return tmpdir
+
+    def test_use_finds_subs(self):
+        """Test that ops.lib.use("baz") works when baz is inside a package in the python path."""
+        tmpdir = self._mkdtemp()
+        sys.path = [tmpdir]
+
+        _mklib(tmpdir, "foo", "bar").write_text(dedent("""
+        LIBNAME = "baz"
+        LIBAPI = 2
+        LIBPATCH = 42
+        LIBAUTHOR = "alice@example.com"
+        """))
+
+        # autoimport would be done in main
+        ops.lib.autoimport()
+
+        # ops.lib.use done by charm author
+        baz = ops.lib.use('baz', 2, 'alice@example.com')
+        self.assertEqual(baz.LIBNAME, 'baz')
+        self.assertEqual(baz.LIBAPI, 2)
+        self.assertEqual(baz.LIBPATCH, 42)
+        self.assertEqual(baz.LIBAUTHOR, 'alice@example.com')
+
+    def test_use_finds_best_same_toplevel(self):
+        """Test that ops.lib.use("baz") works when there are two baz."""
+
+        T = namedtuple("T", "desc sameTop pkgA libA patchA pkgB libB patchB")
+
+        for t in [
+            T("same toplevel, different package, same lib, AB",
+              True, "fooA", "bar", 40, "fooB", "bar", 42),
+            T("same toplevel, different package, same lib, BA",
+              True, "fooA", "bar", 42, "fooB", "bar", 40),
+            T("same toplevel, same package, different lib, AB",
+              True, "foo", "barA", 40, "foo", "barB", 42),
+            T("same toplevel, same package, different lib, BA",
+              True, "foo", "barA", 42, "foo", "barB", 40),
+
+            T("different toplevel, same package, same lib, AB",
+              False, "foo", "bar", 40, "foo", "bar", 42),
+            T("different toplevel, same package, same lib, BA",
+              False, "foo", "bar", 42, "foo", "bar", 40),
+
+            T("different toplevel, different package, same lib, AB",
+              False, "fooA", "bar", 40, "fooB", "bar", 42),
+            T("different toplevel, different package, same lib, BA",
+              False, "fooA", "bar", 42, "fooB", "bar", 40),
+            T("different toplevel, same package, different lib, AB",
+              False, "foo", "barA", 40, "foo", "barB", 42),
+            T("different toplevel, same package, different lib, BA",
+              False, "foo", "barA", 42, "foo", "barB", 40),
+        ]:
+            with self.subTest(t.desc):
+                tmpdirA = self._mkdtemp()
+                sys.path = [tmpdirA]
```

Comment:
setting sys.path directly in a test without a try/finally or a addCleanup trap elsewhere seems dangerous

I guess the whole class has "@patch('sys.path', new=())" around it.

Reply:
yep, that's what that patch is doing :-)